File: ufc_sim_project_streamlit_multi/Assets/Scripts/Core/Math/Formulas.cs
using System;

namespace UFC.Core.Math
{
    public static class Formulas
    {
        public static float Clamp(float x, float lo, float hi)
        {
            return System.Math.Max(lo, System.Math.Min(hi, x));
        }

        public static float AgeBonus(int age)
        {
            return System.Math.Max(-40f, 20f - 4f * System.Math.Abs(age - 30));
        }

        public static float StreakBonus(int streak)
        {
            var s = (int)Clamp(streak, -3, 5);
            return s * 10f;
        }

        public static float EloProb(float ra, float rb)
        {
            return 1f / (1f + (float)System.Math.Pow(10f, (rb - ra) / 400f));
        }

        public static float EffectiveRating(float baseRating, int age, int streak, float sigma, System.Random rng)
        {
            return baseRating + AgeBonus(age) + StreakBonus(streak) + NextGaussian(rng, 0f, sigma);
        }

        private static float RankFactor(int? rankA, int? rankB)
        {
            if (!rankA.HasValue && !rankB.HasValue)
            {
                return 0f;
            }
            float gap = (!rankA.HasValue || !rankB.HasValue) ? 12f : System.Math.Abs(rankA.Value - rankB.Value);
            return System.Math.Min(gap / 15f, 1f) * 0.35f;
        }

        private static float MismatchFactor(float ra, float rb)
        {
            float diff = System.Math.Abs(ra - rb);
            return 0.65f + System.Math.Min(diff / 350f, 1f) * 0.75f;
        }

        private static float UpsetBonus(float expected)
        {
            return System.Math.Max(0f, 0.5f - expected) * 1.6f;
        }

        public static (float, float) ApplyElo(float ra, float rb, bool winnerA, float k = 24f, int? rankA = null, int? rankB = null)
        {
            float pa = EloProb(ra, rb);
            float sa = winnerA ? 1f : 0f;
            float mult = MismatchFactor(ra, rb) + RankFactor(rankA, rankB) + UpsetBonus(winnerA ? pa : (1f - pa));
            mult = Clamp(mult, 0.55f, 2.2f);
            float kEff = k * mult;
            float ra2 = ra + kEff * (sa - pa);
            float rb2 = rb + kEff * ((1f - sa) - (1f - pa));
            return (ra2, rb2);
        }

        private static float NextGaussian(System.Random rng, float mean, float stdDev)
        {
            double u1 = 1.0 - rng.NextDouble();
            double u2 = 1.0 - rng.NextDouble();
            double randStdNormal = System.Math.Sqrt(-2.0 * System.Math.Log(u1)) *
                                   System.Math.Sin(2.0 * System.Math.PI * u2);
            return (float)(mean + stdDev * randStdNormal);
        }
    }
}

File: ufc_sim_project_streamlit_multi/engine/formulas.py
"""Core math formulas for UFC console sim (Flyweight MVP)."""
from __future__ import annotations
import random

def clamp(x: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, x))

def age_bonus(age: int) -> float:
    """Peak around 30; soft decline."""
    return max(-40.0, 20.0 - 4.0 * abs(age - 30))

def streak_bonus(streak: int) -> float:
    s = int(clamp(streak, -3, 5))
    return float(s) * 10.0

def elo_prob(ra: float, rb: float) -> float:
    return 1.0 / (1.0 + 10.0 ** ((rb - ra) / 400.0))

def effective_rating(base_rating: float, age: int, streak: int, sigma: float) -> float:
    return base_rating + age_bonus(age) + streak_bonus(streak) + random.gauss(0.0, sigma)

def _rank_factor(rank_a: int | None, rank_b: int | None) -> float:
    if rank_a is None and rank_b is None:
        return 0.0
    if rank_a is None or rank_b is None:
        gap = 12
    else:
        gap = abs(rank_a - rank_b)
    return min(gap / 15.0, 1.0) * 0.35

def _mismatch_factor(ra: float, rb: float) -> float:
    diff = abs(ra - rb)
    return 0.65 + min(diff / 350.0, 1.0) * 0.75

def _upset_bonus(expected: float) -> float:
    return max(0.0, 0.5 - expected) * 1.6

def apply_elo(
    ra: float,
    rb: float,
    winner_a: bool,
    k: float = 24.0,
    rank_a: int | None = None,
    rank_b: int | None = None,
) -> tuple[float, float]:
    pa = elo_prob(ra, rb)
    sa = 1.0 if winner_a else 0.0
    mult = _mismatch_factor(ra, rb) + _rank_factor(rank_a, rank_b) + _upset_bonus(pa if winner_a else (1.0 - pa))
    mult = clamp(mult, 0.55, 2.2)
    k_eff = k * mult
    ra2 = ra + k_eff * (sa - pa)
    rb2 = rb + k_eff * ((1.0 - sa) - (1.0 - pa))
    return ra2, rb2
